### ISSUE TITLE
Implement DATS/DNA interoperability, Python/R analytical integrations, and portable analysis exchange

### ISSUE THREAD

**Issue by TomiToivio** [TomiToivio/LaclauGPT-Data-Analysis#57]
## Parent architecture

Implement the **Analysis-side** work from `TomiToivio/LaclauGPT#35`.

This repository is the main home for interoperability with **DATS analytical objects**, **DNA/rDNA discourse networks**, and scientific Python/R analysis libraries.

## Goal

Make LaclauGPT Data Analysis able to import external coding/analysis results, run compatible Python/R methods, and export LaclauGPT results back into DATS/DNA-compatible representations without collapsing theoretical distinctions.

The plugin runtime remains authoritative:

```text
CanonicalRecord
 -> generic preprocess
 -> one or more analysis plugins
 -> namespaced AnalyticalResult
 -> generic validation/export
```

## 1. DATS analytical compatibility

Implement a DATS adapter layer for project/document analytical exports.

Map at least:

```text
DATS document              <-> CanonicalRecord external identity
DATS code                  <-> Codebook concept
DATS annotation            <-> EvidenceSpan + Annotation
DATS hierarchy/taxonomy    <-> Codebook hierarchy
DATS AI classification     <-> analysis.plugin_results[dats/...]
DATS human correction      <-> HumanReview
DATS memo/logbook          <-> ResearchNote
DATS concept-over-time     <-> TemporalSeries
DATS whiteboard relation   <-> typed relation / GraphProjection
```

Requirements:

- preserve DATS IDs alongside canonical IDs
- preserve annotation spans exactly
- preserve document/code hierarchy
- preserve coder/model identity
- distinguish human vs machine-generated coding
- preserve confidence/uncertainty where present
- multilingual text support
- multimodal references where present
- CSV plus structured JSON/JSONL/ZIP project import/export where DATS supports them
- synthetic round-trip fixtures

### Optional LLM analysis with DATS data

Support an opt-in workflow where:

1. DATS codes/annotations become context/examples for a LaclauGPT plugin or LLM task.
2. LaclauGPT proposes new annotations/classifications.
3. Proposals retain prompt/model/version/evidence/provenance.
4. Exported proposals are explicitly marked as provisional AI-generated annotations.
5. Human DATS corrections can be re-imported as `HumanReview` and used as reviewed examples.

Follow the methodological pattern of DATS/COTA and Annotation Assistant rather than treating zero-shot LLM output as final coding.

## 2. Define and implement `DiscourseStatement` for DNA/rDNA

Create a versioned analytical object that can bridge LaclauGPT and Discourse Network Analyzer.

Minimum fields:

- stable statement ID
- actor/person/organization ID + label
- concept/claim/demand ID + label
- qualifier / agreement-disagreement / stance where applicable
- document/source ID
- evidence span/text reference
- timestamp
- actor/document attributes
- coder/model
- confidence/uncertainty
- provenance
- human-review status
- external IDs (including DNA IDs)

The object should remain theory-neutral enough to represent DNA coding while allowing LaclauGPT-specific interpretation to point back to it.

## 3. DNA/rDNA import/export

Implement adapters for the DNA ecosystem.

Support as applicable:

- DNA statement-table/project exports -> `DiscourseStatement`
- `DiscourseStatement` -> DNA-compatible statement export
- actor-concept affiliation matrices
- agreement/congruence networks
- disagreement/conflict networks
- longitudinal/windowed discourse networks
- GraphML
- node/edge CSV
- rDNA-readable tabular formats
- rDNA result import as namespaced plugin results

Preserve network construction parameters and projection semantics so a graph remains scientifically interpretable after export/import.

### Optional LLM interpretation of DNA data

Permit an **optional** LaclauGPT step:

```text
DNA-coded statements / networks
 -> descriptive network results
 -> evidence retrieval
 -> LLM-assisted Laclau/Mouffe/Palonen interpretation proposal
 -> human review
```

Examples include candidate articulations, equivalential chains, antagonistic frontiers, nodal/floating signifiers and formations.

Do **not** mechanically infer these concepts from graph metrics.

Important safeguards:

- actor-concept agreement != equivalence
- graph degree != nodal status
- community != ideological formation
- negative edge != antagonism
- network centrality != hegemony

## 4. Python analytical ecosystem integration

Python is the primary runtime. Build thin plugins/adapters around mature scientific libraries rather than duplicating them.

### NLP / representations

- spaCy
- Hugging Face Transformers
- SentenceTransformers
- scikit-learn
- gensim where useful as a baseline

### Topic models

- BERTopic
- classical LDA/NMF baselines
- contextualized topic-model implementations where feasible, especially multilingual/cross-lingual models

### Networks

- NetworkX for convenient graph construction/interchange
- python-igraph for larger analyses
- Leiden community detection (`leidenalg`/igraph where appropriate)

### Statistics

- scipy
- statsmodels
- pandas/polars as appropriate for analysis tables
- PyArrow/Parquet for typed exchange

### Multimodal

Keep multimodal backends optional and plugin-scoped, preserving model/version and evidence references.

## 5. R analysis interoperability under `scripts/r/`

R integrations must be standalone reproducible scripts so researchers can run them in RStudio, `Rscript`, Slurm/batch jobs or another R environment.

Create a documented `scripts/r/` integration area, initially covering:

### quanteda

- canonical records -> quanteda corpus/docvars
- token/dfm/dictionary/KWIC workflows
- export resulting feature/statistical tables back as JSON/CSV/Parquet namespaced analysis outputs

### stm

- canonical corpus + document metadata -> Structural Topic Model
- preserve formula/covariates, seed and model settings
- export topic proportions, prevalence/content effects, correlations and uncertainty

### rDNA

- export `DiscourseStatement`/DNA-compatible tables
- run supported network/congruence/conflict/temporal analyses
- re-import results into namespaced plugin outputs

### igraph

- GraphML/node-edge import/export
- network measures and community results with parameters/provenance

### supporting R packages

- arrow
- data.table/tidyverse where useful
- ggplot2 for static analytical artifacts

Do not require R for the core Python pipeline. R scripts are optional interoperable analytical modules.

## 6. Cross-language interchange

Prefer:

- canonical JSON/JSONL for nested records
- Parquet/Apache Arrow for typed Python<->R tables
- CSV as manual fallback
- GraphML for graphs
- node/edge CSV fallback
- JSON for plugin-result manifests

Avoid opaque pickle/RDS-only handoffs.

Each imported result must record:

- producing tool/library
- version
- parameters/configuration
- source record IDs
- timestamp/run ID
- transformation provenance
- uncertainty where applicable

## 7. Scientific/methodological compatibility documentation

Add an analysis-method matrix linking:

`research question -> theoretical construct -> descriptive method -> software -> output object -> limitations -> validation procedure`

At minimum cover:

- human coding / codebooks
- DATS/COTA concept tracking
- supervised/few-shot classification
- embeddings/similarity
- NER/entity resolution
- BERTopic
- contextualized topic models
- Structural Topic Model
- discourse network analysis
- social/network analysis
- Leiden/community detection
- framing
- temporal trend/change analysis
- correspondence/MCA/GDA where present in the project
- LLM-assisted structured coding and close reading

Explicitly keep **descriptive computational measures separate from discourse-theoretical interpretation**.

## 8. Scientific sources / reading list

Add Analysis-specific references and link to the canonical bibliography in `TomiToivio/LaclauGPT#35`.

Include at minimum:

### General text-as-data

- Grimmer & Stewart (2013), "Text as Data: The Promise and Pitfalls of Automatic Content Analysis Methods for Political Texts", *Political Analysis*.

### DATS / human-in-the-loop

- DATS system paper/documentation.
- COTA / Concept-over-Time, NAACL Demo 2024.
- DATS Annotation Assistant, NAACL Demo 2025.
- DATS Whiteboards / qualitative visual workspace, LREC-COLING 2024.

### DNA

- Philip Leifeld's methodological work on Discourse Network Analysis and dynamic actor-concept networks.
- DNA/rDNA software methodology/publications.

### Text representations/topic models

- Reimers & Gurevych (2019), Sentence-BERT.
- Grootendorst (2022), BERTopic.
- Bianchi, Terragni & Hovy (2021), Contextualized Topic Models.
- Roberts, Stewart, Tingley et al., Structural Topic Model papers / JSS article.
- Benoit et al. (2018), quanteda, *JOSS*.

### Networks

- Traag, Waltman & van Eck (2019), Leiden algorithm, *Scientific Reports*.
- Newman, *Networks*.

### Validation / computational social science

- relevant literature on supervised text classification validation, intercoder reliability, uncertainty and human-in-the-loop annotation.

## 9. Tests

Add synthetic/offline tests for:

- DATS project/document/code/annotation import
- DATS provisional-AI export marking
- DATS human correction re-import
- DNA statement import/export
- GraphML round trip
- Python<->R Parquet/Arrow fixtures
- stable IDs
- Unicode/multilingual text
- timestamp handling
- uncertainty/provenance preservation
- failure on ambiguous/lossy conversion where required fields cannot be safely reconstructed

For R, provide small deterministic fixture scripts/tests that can be run separately from normal Python CI if R is not installed.

## Acceptance criteria

- [ ] Versioned `DiscourseStatement` object implemented/documented.
- [ ] DATS code/annotation/project adapter implemented.
- [ ] DATS optional LLM proposal/review round-trip documented and tested.
- [ ] DNA/rDNA statement import/export implemented.
- [ ] DNA graph/network projections preserve construction semantics.
- [ ] Optional DNA -> LaclauGPT LLM interpretation keeps descriptive/theoretical layers separate.
- [ ] Python integrations exposed through thin analysis plugins/adapters.
- [ ] `scripts/r/` added with quanteda, stm, rDNA and igraph interoperability examples/workflows.
- [ ] Arrow/Parquet + GraphML + JSON/CSV interchange documented.
- [ ] Analysis method/software compatibility matrix added.
- [ ] Scientific sources/readings added.
- [ ] Synthetic cross-tool compatibility tests added.

## Non-goals

- Reimplementing DATS or DNA user interfaces.
- Treating community detection, topic models or embeddings as direct measurements of Laclauian concepts.
- Requiring R for ordinary Python execution.
- Storing private research corpora or real DATS/DNA projects in public fixtures.

**Comment by TomiToivio** [comment]
Cross-repository links:

- Canonical theory/method/data contract: https://github.com/TomiToivio/LaclauGPT/issues/35
- Collection interoperability: https://github.com/TomiToivio/LaclauGPT-Data-Collection/issues/65
- Visualization interoperability: https://github.com/TomiToivio/LaclauGPT-Data-Visualization/issues/30

Analysis owns the theory-neutral `DiscourseStatement` / analytical interchange objects and DATS/DNA conversion logic; Visualization should consume those objects rather than reconstructing analysis from rendered graphs.

**Comment by TomiToivio** [comment]
Implemented #57 on `main`.

Key pieces now in place:

- versioned, theory-neutral `DiscourseStatement` plus DATS/DNA exchange contracts
- DATS project/code/annotation import/export with exact spans, stable external IDs, machine-vs-human status, provisional AI export marking, and human correction/review re-import
- DNA/rDNA statement-table/native interoperability plus graph/network exchange semantics
- portable GraphML and node/edge exchange with required projection/construction metadata
- optional Arrow/Parquet handoff with an explicit LaclauGPT result manifest
- namespaced external/plugin results kept `DESCRIPTIVE_ONLY` unless a separate evidence-grounded interpretive stage is reviewed
- existing thin Python backends for spaCy, Transformers, SentenceTransformers, sklearn, BERTopic, gensim, statsmodels, NetworkX plus optional igraph/Leiden and columnar dependencies
- optional `scripts/r/` workflows for quanteda, STM, rDNA and igraph; R remains outside the core runtime
- analysis-method/software compatibility matrix, safeguards, and scientific references in `docs/INTEROPERABILITY_ANALYSIS.md`
- synthetic DATS and DNA fixtures plus offline tests for Unicode/spans, stable IDs, provisional AI status, human review, DNA CSV round trips, graph semantics and lossy-conversion failure

I also fixed a final integration problem discovered during review: the repo had both `laclaugpt_data_analysis/interoperability.py` and an `interoperability/` package. Python resolves the package first, so the module API could be shadowed. Commit `cc20a812359f386270ffb0b2514c8f9e009ddce3` moves the contracts under the package, exports the stable public API from `interoperability/__init__.py`, and removes the colliding module while preserving the native `interoperability.dna` adapters.

No GitHub Actions run was attached to the latest main commit when checked, so I am not claiming a remote CI pass; the requested synthetic/offline tests are present in the repository.

**Comment by TomiToivio** [comment]
## Final validation

Follow-up validation is now complete on `main`.

GitHub Actions run **#225** (`35205693439`) is green across the full configured matrix:

- lint changed Python files: **pass**
- Python 3.11 tests: **pass**
- Python 3.12 tests: **pass**
- Python 3.13 tests: **pass**

During validation I fixed two integration details that were not visible when the issue was first closed:

1. The interoperability contracts had been consolidated into the `laclaugpt_data_analysis.interoperability` package, so package-local compatibility imports for canonical models had to be restored to avoid collection-time import errors.
2. `tests/test_legacy_discourse_parity.py` still asserted the old literal prompt wording/order. The current prompt library deliberately renders `[CONTEXT: CODEBOOK CANDIDATES, NOT EVIDENCE]` through the canonical context envelope, so the test now checks the current semantic contract and verifies `(none available)` when no codebook is supplied.

The implementation and its regression suite are therefore validated against the repository's normal CI rather than only by static review.
